src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java

```diff
@@ -92,4 +137,18 @@ public void close() {
       myWorkerFuture.cancel(true);
     }
   }
+
+  /**
+   * Closes the state object, and blocks until all pending async work is finished. The state object
+   * will reset to a clean slate after this method finishes. This may only be called from the host
+   * Skyframe thread.
+   */
+  public void closeAndWaitForTermination() throws InterruptedException {
+    close();
+    workerExecutorService.close(); // This blocks
+    reset();
```

Comment:
Why is this call necessary here? AFAIU after this is called, the only thing that could possibly ever happen to `RepoFetchingSkyKeyComputeState` is another `close()` call.

In addition the aesthetic issue, this creates an unnecessary `ExecutionService`.

Reply:
in the case of test_keep_going_weird_deadlock, the SkyFunction is called a second time. Similar to error bubbling, IIUC.

Reply:
Huh, remind me again, why is the SkyFunction called a second time in `test_keep_going_weird_deadlock`?

Reply:
because the host Skyframe thread returns `null`, but the worker has already died (in fact, the worker itself returned a `null` in the Future). Normally the worker should never return `null` -- on `valuesMissing` it would block for a fresh env and try again; but in this case, even the fresh env still returns `null` for a `getValue` call, so the worker just returns that `null`.

Reply:
Thanks! Please add a comment to explain this scenario. If I haven't figured this out by myself after staring at this problem for so much time, people coming in fresh have no chance.

Reply:
done.